app/src/main/java/com/amaze/filemanager/filesystem/ssh/Statvfs.java

```diff
@@ -0,0 +1,179 @@
+ */
+
+package com.amaze.filemanager.filesystem.ssh;
+
+import net.schmizz.sshj.common.Buffer;
+import net.schmizz.sshj.sftp.PacketType;
+import net.schmizz.sshj.sftp.Request;
+import net.schmizz.sshj.sftp.SFTPClient;
+import net.schmizz.sshj.sftp.SFTPException;
+
+import java.math.BigInteger;
+
+/**
+ * Wrapper for SSH <code>statvfs@openssh.com</code> request/response.
+ *
+ * This is not specified in official SSH protocol; it is implemented by OpenSSH (which is the most
+ * used SSH server implementation around.
+ */
+public class Statvfs
+{
+    private static final String STATVFS_OPENSSH_COM = "statvfs@openssh.com";
+
+    /**
+     * Convenience method for creating <code>statvfs@openssh.com</code> {@link Request}s.
+     *
+     * @param sftpClient {@link SFTPClient} instance
+     * @param path remote path
+     * @return {@link Request}
+     */
+    public static final Request request(final SFTPClient sftpClient, final String path){
+        return sftpClient.getSFTPEngine().newRequest(PacketType.EXTENDED)
+                .putString(STATVFS_OPENSSH_COM)
+                .putString(path,
+                        sftpClient.getSFTPEngine().getSubsystem().getRemoteCharset());
+    }
+
+    /**
+     * Wrapper object for {@link net.schmizz.sshj.sftp.Response}.
+     *
+     * PROTOCOL specification defines the packet structure as
+     *
+     * <code>
+     * uint64          f_bsize         // file system block size
+     * uint64          f_frsize        // fundamental fs block size
+     * uint64          f_blocks        // number of blocks (unit f_frsize)
+     * uint64          f_bfree         // free blocks in file system
+     * uint64          f_bavail        // free blocks for non-root
+     * uint64          f_files         // total file inodes
+     * uint64          f_ffree         // free file inodes
+     * uint64          f_favail        // free file inodes for to non-root
+     * uint64          f_fsid          // file system id
+     * uint64          f_flag          // bit mask of f_flag values
+     * uint64          f_namemax       // maximum filename length
+     * </code>
+     *
+     * whereas <code>f_flag</code> is defined as
+     *
+     * <code>
+     * #define SSH_FXE_STATVFS_ST_RDONLY       0x1     // read-only
+     * #define SSH_FXE_STATVFS_ST_NOSUID       0x2     // no setuid
+     * </code>
+     *
+     * @see <a href="https://cvsweb.openbsd.org/cgi-bin/cvsweb/src/usr.bin/ssh/PROTOCOL?annotate=HEAD">PROTOCOL specification</a>
+     */
+    public static class Response
+    {
+        public final net.schmizz.sshj.sftp.Response mResponse;
+
+        /*
+            Regarding choice of number types:
+
+            - fileSystemFlag is only 0x0, 0x1 and 0x2
+            - filenameMaxLength should be well within Long.VALUE_MAX
+
+            Others, can use BigInteger to prepare for the ever increasing disk size for years to come.
+         */
```

Comment:
Actually several of those container types are hilariously oversized.

Maximum length of filename (a component of file path between two slashes) is directly tied to size of the memory buffer, used for listing files in directory. I strongly suspect, that we won't *need* filenames longer than 2Gb for couple hundred years at least. Even if we somehow start using such names, an Android file manager won't be able to allocate an array for storing those names, because maximum length of Java array is slightly below `Integer.MAX_VALUE`.

File system block size is another example — if file systems with such ridiculously huge blocks become commonplace, knowing the size of block will be useless, because using temporary buffers of such sizes is going to be impractical. May as well cap them to 2Gb.

By the way, it is not necessary to use `BigInteger` for storing unsigned 64-bit values — `long` can store them fine. If necessary, you can allocate an instance of `BigInteger` lazily before actually operating on it.

Reply:
Thanks for the suggestion.

ffca0bb changes them. Also changed `fileSystemFlag` to `Integer` as valid values are actually 0, 1, 2 (per specification) or 3 even if `|` flag is applied.

However, OpenSSH on Linux (at least on my Xenial machines) still requires `BigInteger`. Sorry about this.

Reply:
@TranceLove It is largely inconsequential (really, how many time would that method be called during a copy? one?), but let's clarify that point: since you are calling `readUInt64AsBigInteger`, it means, that the size of value should be the same as `long` — 64 bits. The relevant methods in jssh seem to read exactly same amount of bytes from socket: [1][1], [2][2]. So there can't be any objective reason, requiring use of BigInteger.

The only difference between long and BigInteger is treatment of sign bit: later preserves original unsigned value, former may "overflow" into negative for sufficiently high filesystem IDs. But you will only notice the difference if you *print the data to user* or *perform math on it*. There is nothing wrong with simply *storing* unsigned integer in signed Java data type and [converting to BigInteger before printing it to user](https://stackoverflow.com/questions/7031198).

Btw, based on the style of your code, it looks like you have some experience programming for .Net or other "big" managed platforms. In most of those platforms using object-like data types for primitives is preferable (or the only) choice. In Java primitive scalar types (`long`, `int`, 'char') are more widely used for that purpose. There is zero benefit to using `Long.valueOf` and taking `intValue` from it the way you do. Just cast to int:

````java
filenameMaxLength = (int) mResponse.readUInt64();
````

[1]: https://github.com/hierynomus/sshj/blob/8ca6451d5dee2667d1a5/src/main/java/net/schmizz/sshj/common/Buffer.java#L351
[2]: https://github.com/hierynomus/sshj/blob/8ca6451d5dee2667d1a5/src/main/java/net/schmizz/sshj/common/Buffer.java#L343

Reply:
@TranceLove Of course the check behind your exception is meaningful — as I explained above, it might be bad idea to expose an unsigned value stored in `long`, because it may overflow when you print it out or operate on it (as it did in your example). But the check is mainly there to warn you about dangers of overflow. If you are already conscious of that, you can make that part of contract.

Personally I would do it like this:

1. Make `fileSystemId` long
2. Make `fileSystemId` private
3. Create two getters: `BigInteger getFsId()` and `long getRawFsId()`
4. `toString()` shall print result of the first getter

This will avoid creating and storing instance of `BigInteger` unless somebody actually uses the field.

Old versions of Java used to have no good utility methods for working with unsigned values. But since Java 8, there are multiple static methods in `Long` class (such as [toUnsignedString](https://docs.oracle.com/javase/8/docs/api/java/lang/Long.html#toUnsignedString-long-)), that allow you to treat scalar `long` values as unsigned. If you need to work with 64-bit unsigned numbers, BigInteger is not a must anymore.

As for the silly check, here is relevant source code:

````java
public long readUInt64()
            throws BufferException {
        long uint64 = (readUInt32() << 32) + (readUInt32() & 0xffffffffL);
        if (uint64 < 0) // this line has tripped you
            throw new BufferException("Cannot handle values > Long.MAX_VALUE");
        return uint64;
}
````

You can just read the value yourself without checking for negativity:

````java
public long readUInt64FromBuffer(net.schmizz.sshj.common.Buffer buffer) {
        long uint64 = (buffer.readUInt32() << 32) + (buffer.readUInt32() & 0xffffffffL);
        return uint64;
}
````

Reply:
Excellent. Never looked that deep into sshj's source code - thanks for pointing out!